### PR TITLE
2020 1 11 accounts

### PR DESCRIPTION
**Pull request by Christewart**
This is built on top of #1010 .

Basically this further integrates `HDAccount` into the wallet. It gives us the ability to 

- Query the balance of a specific `HDAccount`
- Fund a transaction from a specific `HDAccount`
- Generate an address for a specific `HDAccount`

This is needed for multi-user wallets, or wallets that want to segregate funds based on application.

### CODE REVIEW

File: docs/applications/wallet.md [hunk 1]
```diff
@@ -62,7 +62,7 @@ implicit val walletConfig = WalletAppConfig(datadir, config)
 import org.bitcoins.chain.config.ChainAppConfig
 implicit val chainConfig = ChainAppConfig(datadir, config)
 
-// when this future completes, we have
+// when this future completes, we havehttps://en.wikipedia.org/wiki/Liveness
```
Comment (benthecarman) [inline docs/applications/wallet.md]: accidental paste

File: core/src/main/scala/org/bitcoins/core/hd/HDAccount.scala [hunk 2]
```diff
@@ -23,3 +25,39 @@ case class HDAccount(
   def toChain(chainType: HDChainType): HDChain =
     HDChain(chainType = chainType, account = this)
 }
+
+object HDAccount {
+
+  /** This method is meant to take in an arbtirary bip32 path and see
```
Comment (benthecarman) [inline core/src/main/scala/org/bitcoins/core/hd/HDAccount.scala]: > arbtirary

arbitrary

File: core/src/main/scala/org/bitcoins/core/hd/HDAccount.scala [hunk 3]
```diff
@@ -23,3 +25,39 @@ case class HDAccount(
...
+    * the same account
+    * */
+  def isSameAccount(path: Vector[BIP32Node], account: HDAccount): Boolean = {
+    if (account.path.length > path.length) {
+      false
+    } else {
+      val zipped = path.zip(account.path)
+      zipped.foldLeft(true) {
+        case (past, (bip32Node, accountNode)) =>
+          if (past) {
```
Comment (benthecarman) [inline core/src/main/scala/org/bitcoins/core/hd/HDAccount.scala]: can simplify to `past && bip32Node == accountNode`

File: core/src/main/scala/org/bitcoins/core/hd/HDAccount.scala [hunk 4]
```diff
@@ -23,3 +25,39 @@ case class HDAccount(
   def toChain(chainType: HDChainType): HDChain =
     HDChain(chainType = chainType, account = this)
 }
+
+object HDAccount {
```
Comment (benthecarman) [inline core/src/main/scala/org/bitcoins/core/hd/HDAccount.scala:29]: I think it'd be useful to have a function that takes a `BIP32Path` as an input and returns a `HDAccount`

Comment (Christewart) [reply core/src/main/scala/org/bitcoins/core/hd/HDAccount.scala:29]: Do you have anything in mind to use it for at the moment?

File: core-test/src/test/scala/org/bitcoins/core/hd/HDAccountTest.scala [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+  }
+
+  val defaultAcct = WalletTestUtil.defaultHdAccount
+  val defaultPath = defaultAcct.path
+  it must "determine if a bip32 path is the same as a given HDAccount" in {
+    val isSameDefault = HDAccount.isSameAccount(defaultPath, defaultAcct)
+    assert(isSameDefault, s"Must have symmetry")
+  }
+
+  it must "fail if the given path is shorter than how BIP44 defines accounts" in {
```
Comment (nkohen) [inline core-test/src/test/scala/org/bitcoins/core/hd/HDAccountTest.scala:22]: should probably also fail if the path isn't hardened in the right places?

File: core/src/main/scala/org/bitcoins/core/hd/HDAccount.scala [hunk 6]
```diff
@@ -23,3 +25,34 @@ case class HDAccount(
...
+    * This is tricky as an account is defined as
+    * m / purpose' / cointype' / account'
+    *
+    * whereas a bip32 path can be arbitrarily deep.
+    *
+    * We want to just check the first 4 elements of the path
+    * and see if they are the same, which indicates we are in
+    * the same account
+    * */
+  def isSameAccount(path: Vector[BIP32Node], account: HDAccount): Boolean = {
```
Comment (nkohen) [inline core/src/main/scala/org/bitcoins/core/hd/HDAccount.scala:43]: nit: `isInSameAccount`?

File: wallet/src/main/scala/org/bitcoins/wallet/api/WalletApi.scala [hunk 7]
```diff
@@ -111,12 +111,27 @@ trait LockedWalletApi extends WalletApi {
     } yield confirmed + unconfirmed
   }
 
+  def getBalance(account: HDAccount): Future[CurrencyUnit] = {
```
Comment (nkohen) [inline wallet/src/main/scala/org/bitcoins/wallet/api/WalletApi.scala:115]: scaladoc

**Review by nkohen — COMMENTED** [review]
Just some small stuff

File: core/src/main/scala/org/bitcoins/core/hd/HDAccount.scala (hunk 6 above)
Comment (rorp) [reply core/src/main/scala/org/bitcoins/core/hd/HDAccount.scala:43]: Maybe `HDAccount.contains(path: Vector[BIP32Node])`?

File: core/src/main/scala/org/bitcoins/core/hd/HDAccount.scala (hunk 4 above)
Comment (benthecarman) [reply core/src/main/scala/org/bitcoins/core/hd/HDAccount.scala:29]: I have a hww, and only want to be able to spend from account `2`, when given a `BIP32Path` I want to verify that it is the correct `HDAccount`

File: testkit/src/main/scala/org/bitcoins/testkit/wallet/FundWalletUtil.scala [hunk 8]
```diff
@@ -0,0 +1,117 @@
...
+  def fundAccountForWallet(
+      amts: Vector[CurrencyUnit],
+      account: HDAccount,
+      wallet: Wallet)(implicit ec: ExecutionContext): Future[Wallet] = {
+
+    val addressesF = Future.sequence(Vector.fill(3) {
+      //this Thread.sleep is needed because of
+      //https://github.com/bitcoin-s/bitcoin-s/issues/1009
+      //once that is resolved we should be able to remove this
+      Thread.sleep(500)
```
Comment (rorp) [inline testkit/src/main/scala/org/bitcoins/testkit/wallet/FundWalletUtil.scala]: `1 to 3 foldLeft (Future.successful(Vector())) (... getNewAddress ...)` works every time without `Thread.sleep()`.

**Pull request merged**